### ISSUE TITLE
Meaningful SensorIDs

### ISSUE THREAD

**Issue by nbarlowATI** [alan-turing-institute/CROP#224]
(Somewhat related to #215 )
We currently mostly use an integer primary key for SensorID, which basically just represents the order that sensors were put into our database.
We also have, for the Aranet Temperature and Relative Humidity sensors that we currently get data via Zensie for, a "Device ID", which is a long string representing the URL for that sensor's data on the Zensie platform.

More useful would be something that is actually written on the sensors themselves, so that the people at the farm can also recognise it.
The spreadsheet we got from Serena has a couple of potentially relevant columns:
* **Aranet Pro ID**: this is an integer, with a value in the spreadsheet for all-but-one sensor (an air velocity one).
* **Serial Number**: this seems to be a long integer, with a series of zeros in the middle.   We don't have serial numbers in the spreadsheet for some of the sensors, including some T/RH sensors in the farm.
* **Code**: A hexidecimal number, which is apparently also written on the sensors themselves.   Seems to be related to serial number (the hex representation of the bit after the zeros in the serial number).   We do have this for all sensors in the spreadsheet.

So, it seems it would make sense to add _at least_ the "Code" to our database, perhaps as a new column in the Sensor table?  Perhaps also add Serial Number as a nullable field ? 

Is there a non-disruptive (or minimally disruptive) way of doing this?   
We presumably need to : 
* Dump the current contents of the Sensor and SensorLocation contents, and re-populate them? 
* Change the ForeignKey for all the Readings classes, or at least the logic that associates a reading to a sensor.
* Would we also need to dump/refill all the historical Readings in the db?

**Comment by mhauru** [comment]
Would we be able to avoid the dump/refills if we didn't make the code of the sensor the primary key? So that we would only add a column to the sensors table, and not change anything else.

**Comment by nbarlowATI** [comment]
I guess.... probably?   That would seem to be the easiest solution if so...

**Comment by mhauru** [comment]
Thinking of this now, the sensors table is a bit funny, in that it lists sensors of various types, not just Aranet, and the current `device_id` column holds very different things for different types of sensors.

**Comment by mhauru** [comment]
An even less invasive change would be to add a table called `aranet_codes` that would map current integer device IDs to codes and serial numbers. I personally find these very thin but numerous SQL tables a bit unaesthetical, but I know nothing about database design, and people tell me there are reasons to favour them (what you were saying about James's talk).

**Comment by nbarlowATI** [comment]
I'm also not sure about this, but if we can do this by adding a column to the current table (or repurposing the `device_id` (though maybe we use this to actually get the data from zensie right now?)) I think that would be cleaner - a new table seems like overkill if there is a 1:1 mapping between sensor and aranet_code.   The Sensor table should contain all the attributes that are intrinsically and immutably associated to that sensor (which is why SensorLocation should be a different table given that sensors can be moved).    At least that's my take on it...

**Comment by mhauru** [comment]
Yes, that makes sense. Would you mind if `code` wasn't the primary key, and we still kept our own integer keys?

>(or repurposing the device_id (though maybe we use this to actually get the data from zensie right now?))

Yes, I think that's right. Those device IDs are what the 30MHz platform API needs to identify the sensors.

**Comment by nbarlowATI** [comment]
Yep, happy to keep our own integer key as the primary key, and just add code as a new column... That does seem to be the simplest.

**Comment by mhauru** [comment]
Ok, I can get to adding these for the ones we know.

**Comment by mhauru** [comment]
Database updated for `cropapp` too. Still need to update the Sensor List page columns.
